sycl/doc/extensions/SYCL_pipes_usage.rst

```diff
@@ -0,0 +1,79 @@
+SYCL INTEL spatial pipes
+========================
+
+Introduction
+============
+
+Pipe is a memory object that stores data organized as a FIFO buffer.
+This implementation enables 2 classes of pipe connectivity:
+  Cross kernel: Kernel A -> Kernel B
+  Intra-kernel: Kernel A -> Kernel A
+
+and 2 types of pipes: non-blocking and blocking.
+
+Blocking calls wait until there is available capacity to commit data, or until
+data is available to be read. In all of these cases the pipe indirectly conveys
+side channel control information to the program, which can be used for flow
+control or many other purposes by an application.
+
+Unsuccessful non-blocking pipe reads or writes do not impact the state or
+content of a pipe.
+
+Pipe identification
+===================
+
+Identification mechanism of a pipe is base on template type specialization of
+pipe class.
+
+Consider following code:
+.. code:: cpp
+  template <class name, class dataT, size_t min_capacity = 0> class pipe;
+
+where 'name' is a name of a pipe that helps to identify pipe,
+      'dataT' is a type of data to store into the pipe,
+      'min_capacity' is the number of outstanding words that can be written to
+                     a pipe but not yet read from it.
+
+The combined set of the three template parameters forms the type of a pipe.
+Any uses of a read/write method on that type operate on the same pipe by
+static connectivity of pipes.
+
+Interaction with SYCL API
+=========================
+
+Following methods of pipe class are available for user:
+.. code:: cpp
+  // Non-blocking pipes
+  static dataT read(bool &success_code);
+  static void write(dataT data, bool &success_code);
+
+  // Blocking pipes
+  static dataT read();
+  static void write(dataT data);
+
+Writes to or reads from a pipe are accesses to a pipe with the same pipe type.
+
+Simple example of SYCL program
+==============================
+.. code:: cpp
+  using Pipe = pipe<class some_pipe, int, 1>;
+  Queue.submit([&](handler& cgh) {
+    cgh.parallel_for<class foo_nb>(range<1> { dataSize }, [=](id<1> idx) {
+      bool SuccessCode = false;
+      while (!SuccessCode)
+        Pipe::write(42, SuccessCode); // Write '42' into a some_pipe allocated
```

Comment:
What is the memory model of a `pipe` in the case you are executing it in a `parallel_for` and not a `single_task`?

Reply:
Do you expect some independent-forward-progress here in the case `dataSize > 1` (the capacity of the pipe...)?

Reply:
Yes, I expect that. I should re-write the example code into:
`cgh.parallel_for<class bar>(range<1> { 1024 }, [=](id<1> idx) {
    write_acc[idx] = my_pipe::read();
  });`

OpenCL analog of the code above is:
`__kernel void foo(__read_only pipe uint in_pipe, __global uint *dst) {
    int gid = get_global_id(0);
    read_pipe(in_pipe, &dst[gid]);
}`

Reply:
But it is illegal OpenCL code in general. You have to use pipe reservation for this and and I am not sure we want this anyway on FPGA...
So the assumptions are probably to be described in the document describing the extension.
What happens if there are multiple work-group on different CU using the same pipe input or output?

Reply:
> But it is illegal OpenCL code in general. You have to use pipe reservation for this and and I am not sure we want this anyway on FPGA...
You are right.

> What happens if there are multiple work-group on different CU using the same pipe input or output?
... to be described in the document describing the extension :) I added a quote from there:

**Accesses to a single pipe within two work-items of the same kernel, and/or threads of the host program:** No ordering guarantees are made on the order of pipe operations across device work-items or host threads. For example, if two work-items executing a kernel write to a pipe, there are no guarantees that the work-item with lower *id* (for any definition of *id*) executes before the pipe write from a higher *id*. The execution order of work-items executing a kernel are not defined by SYCL, may be dynamically reordered, and may not be deterministic. If ordering guarantees are required across work-items and/or host threads, synchronization mechanisms such as atomics or barriers must be used.

Reply:
> @keryell, are you okay to merge this PR?

Yes I am fine with merging.
It looks like in https://github.com/intel/llvm/pull/318 there are some corrections that will be done later.